File: scripts/alert-monitor.js
/**
 * 异常告警监控脚本
 * 监控冲突堆积、入库失败、向量嵌入异常，输出到告警日志文件
 */

const Database = require('better-sqlite3');
const fs = require('fs');
const path = require('path');

const DB_PATH = process.env.CACHE_DB_PATH || path.join(__dirname, '..', 'cache', 'drafts.db');
const ALERT_LOG_PATH = process.env.ALERT_LOG_PATH || path.join(__dirname, '..', 'cache', 'alerts.log');

const CONFLICT_THRESHOLD = parseInt(process.env.CONFLICT_ALERT_THRESHOLD || '10', 10);
const COMMIT_FAIL_THRESHOLD = parseInt(process.env.COMMIT_FAIL_ALERT_THRESHOLD || '5', 10);
const EMBED_FAIL_THRESHOLD = parseInt(process.env.EMBED_FAIL_ALERT_THRESHOLD || '3', 10);

function writeAlert(level, category, message, detail) {
  const timestamp = new Date().toISOString();
  const line = `[${timestamp}] [${level}] [${category}] ${message} | ${JSON.stringify(detail)}\n`;
  fs.appendFileSync(ALERT_LOG_PATH, line);
  console.log(`[alert-monitor] ${line.trim()}`);
}

function checkConflicts(db) {
  const stmt = db.prepare(`SELECT COUNT(*) as count FROM conflicts WHERE resolution IS NULL`);
  const { count } = stmt.get();
  if (count >= CONFLICT_THRESHOLD) {
    writeAlert('WARNING', 'conflict_pileup', `未处理冲突数量 ${count}，超过阈值 ${CONFLICT_THRESHOLD}`, { count });
  }
}

function checkCommitFailures(db) {
  const stmt = db.prepare(`
    SELECT COUNT(*) as count FROM audit_log
    WHERE action = 'commit' AND detail LIKE '%failed%'
      AND created_at > datetime('now', '-1 hour')
  `);
  const { count } = stmt.get();
  if (count >= COMMIT_FAIL_THRESHOLD) {
    writeAlert('ERROR', 'commit_failure', `近 1 小时入库失败 ${count} 次，超过阈值 ${COMMIT_FAIL_THRESHOLD}`, { count });
  }
}

function checkEmbedFailures() {
  // 读取 GBrain 嵌入日志或检查索引状态，这里简化为检查环境变量模拟
  const embedFailCount = parseInt(process.env.EMBED_FAIL_COUNT || '0', 10);
  if (embedFailCount >= EMBED_FAIL_THRESHOLD) {
    writeAlert('ERROR', 'embed_failure', `向量嵌入失败 ${embedFailCount} 次，超过阈值 ${EMBED_FAIL_THRESHOLD}`, { count: embedFailCount });
  }
}

function runAlertMonitor() {
  const db = new Database(DB_PATH);
  try {
    checkConflicts(db);
    checkCommitFailures(db);
    checkEmbedFailures();
    console.log('[alert-monitor] 监控检查完成');
  } finally {
    db.close();
  }
}

if (require.main === module) {
  runAlertMonitor();
}

module.exports = { runAlertMonitor };

File: scripts/alert_monitor.py
#!/usr/bin/env python3
"""异常告警监控脚本。

监控冲突堆积、入库失败、向量嵌入异常，输出到告警日志文件。
"""

import json
import os
import sqlite3
import sys
from datetime import datetime, timedelta
from pathlib import Path

PROJECT_DIR = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(PROJECT_DIR))

DB_PATH = os.environ.get("CACHE_DB_PATH", str(PROJECT_DIR / "cache" / "drafts.db"))
ALERT_LOG_PATH = os.environ.get("ALERT_LOG_PATH", str(PROJECT_DIR / "cache" / "alerts.log"))

CONFLICT_THRESHOLD = int(os.environ.get("CONFLICT_ALERT_THRESHOLD", "10"))
COMMIT_FAIL_THRESHOLD = int(os.environ.get("COMMIT_FAIL_ALERT_THRESHOLD", "5"))
EMBED_FAIL_THRESHOLD = int(os.environ.get("EMBED_FAIL_ALERT_THRESHOLD", "3"))


def write_alert(level: str, category: str, message: str, detail: dict):
    """写入告警日志。"""
    timestamp = datetime.now().isoformat()
    line = f"[{timestamp}] [{level}] [{category}] {message} | {json.dumps(detail, ensure_ascii=False)}\n"
    with open(ALERT_LOG_PATH, "a", encoding="utf-8") as f:
        f.write(line)
    print(f"[alert-monitor] {line.strip()}")


def check_conflicts(conn: sqlite3.Connection):
    """检查未处理冲突堆积。"""
    cursor = conn.execute("SELECT COUNT(*) as count FROM conflicts WHERE resolution IS NULL")
    count = cursor.fetchone()["count"]
    if count >= CONFLICT_THRESHOLD:
        write_alert(
            "WARNING",
            "conflict_pileup",
            f"未处理冲突数量 {count}，超过阈值 {CONFLICT_THRESHOLD}",
            {"count": count},
        )


def check_commit_failures(conn: sqlite3.Connection):
    """检查近 1 小时入库失败次数。"""
    one_hour_ago = (datetime.now() - timedelta(hours=1)).isoformat()
    cursor = conn.execute(
        "SELECT COUNT(*) as count FROM audit_log "
        "WHERE action = 'commit' AND detail LIKE '%failed%' AND created_at > ?",
        (one_hour_ago,),
    )
    count = cursor.fetchone()["count"]
    if count >= COMMIT_FAIL_THRESHOLD:
        write_alert(
            "ERROR",
            "commit_failure",
            f"近 1 小时入库失败 {count} 次，超过阈值 {COMMIT_FAIL_THRESHOLD}",
            {"count": count},
        )


def check_embed_failures():
    """检查向量嵌入失败次数（简化为环境变量模拟）。"""
    embed_fail_count = int(os.environ.get("EMBED_FAIL_COUNT", "0"))
    if embed_fail_count >= EMBED_FAIL_THRESHOLD:
        write_alert(
            "ERROR",
            "embed_failure",
            f"向量嵌入失败 {embed_fail_count} 次，超过阈值 {EMBED_FAIL_THRESHOLD}",
            {"count": embed_fail_count},
        )


def run_alert_monitor():
    """执行告警监控检查。"""
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    try:
        check_conflicts(conn)
        check_commit_failures(conn)
        check_embed_failures()
        print("[alert-monitor] 监控检查完成")
    finally:
        conn.close()


if __name__ == "__main__":
    run_alert_monitor()
